docs/source/training_vlm_sft.md

```diff
@@ -0,0 +1,386 @@
+)
+
+training_args.remove_unused_columns = False                         # Ensure unused columns are not removed in the collator (important for batch processing).
+
+```
+
+The `collate_fn` is responsible for processing and preparing individual examples to form a batch.  
+
+Each example in the batch undergoes the following steps:  
+
+1. The **chat template** is applied to the text.  
+2. The **processor tokenizes** both `texts` and `images`, encoding them into tensors.  
+3. The **labels** for training are set as the `input_ids` of the example.  
+4. Certain **special tokens** are **masked (ignored)** during loss computation:  
+   - `pad_token_id`  
+   - `<image_token_id>`  
+   - `<image_soft_token>` (corresponding to ID `262144`)  
+
+This process is similar across different dataset types, with a minor variation in how images are handled:  
+
+- **Single Image + Text** → A **list of images** is directly processed.  
+- **Multi-Image + Text** → A **list of lists of images** is used, where each batch element contains multiple images.  
+
+```python
+
+# For multi-image cases
+def process_vision_info(messages: list[dict]) -> list[Image.Image]:
+    image_inputs = []
+    for msg in messages:
+        content = msg.get("content", [])
+        if not isinstance(content, list):
+            content = [content]
+
+        for element in content:
+            if isinstance(element, dict) and ("image" in element or element.get("type") == "image"):
+                if "image" in element:
+                    image = element["image"]
+                else:
+                    image = element
+                if image is not None:
+                    image = Image.open(io.BytesIO(image["bytes"]))
+                    image_inputs.append(image.convert("RGB"))
+    return image_inputs
+
+def collate_fn(examples):
+    texts = [processor.apply_chat_template(example["messages"], tokenize=False) for example in examples]
+    if "images" in examples[0]:  # single-image
+        images = [
+            img.convert("RGB") if img.mode == "RGBA" else img for example in examples for img in example["images"]
+        ]
+    else:  # multi-image
+        images = [process_vision_info(example["messages"]) for example in examples]
+
+    # Tokenize the texts and process the images
+    batch = processor(
+        text=texts, images=images, return_tensors="pt", padding=True
+    )  # Encode texts and images into tensors
+
+    # The labels are the input_ids, and we mask the padding tokens in the loss computation
+    labels = batch["input_ids"].clone()  # Clone input IDs for labels
+    # Mask image tokens
+    image_token_id = [
+        processor.tokenizer.convert_tokens_to_ids(processor.tokenizer.special_tokens_map["boi_token"])
+    ]
+    # Mask tokens for not being used in the loss computation
+    labels[labels == processor.tokenizer.pad_token_id] = -100
```

Comment:
Is the eos different from the pad token? because we we to mask the pad tokens, but not the eos tokens.

Reply:
It is 😄 

![Screenshot 2025-04-09 at 15 19 02](https://github.com/user-attachments/assets/c3a3d909-a0af-4252-a8e5-8c995c4fbf29)